### ISSUE TITLE
[DependencyInjection] env vars are not reset upon kernel.reset

### ISSUE THREAD

**Issue by faizanakram99** [symfony/symfony#59128]
### Symfony version(s) affected

7.1.0

### Description

#54666 was supposed to reset env vars upon kernel.reset but seems it doesn't work as intended, it probably works for web environments with booted kernels like php-pm or frankenphp worker mode as container is reset for them on each kernel boot but not for messenger workers. Currently there are two issues with it.

1. EnvVarProcessor is not tagged with kernel.reset
2. Env vars are cached higher up in stack in Container (service_container) and they are never reset.

See https://github.com/symfony/symfony/pull/54666#issuecomment-2483122814
            

### How to reproduce

Reproducer isn't needed, the description is enough to explain the issue. 

### Possible Solution

First issue can be fixed by adding  `->tag('kernel.reset', ['method' => 'reset'])` [here](https://github.com/symfony/symfony/blob/75da5aaa8254fb9b09cdaaea996d922bd0b89874/src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php#L193)

Second issue, I'm not sure, ~~maybe we should get rid of services resetter altogether and just reboot kernel~~ (edit: bad idea as pointed out by @stof in following comment). Alternatively, add `service_container` to services_resetter but just reset envCache (and not other services). It would require another public method other than reset in container class and tagging it with `kernel.reset` and use that method as "resetMethod". Instead of tagging service_container with kernel.reset, we could simply call that public method from `EnvVarProcessor::reset()`

### Additional Context

Resetting env vars is necessary to be able to run symfony messenger workers for a multi tenant application (single web server multiple databases) where each message is identified with a tenant stamp and depending upon the tenant id, its env var needs are loaded like database_url, filesystem path etc. Also it is similar to web model that way, each message is handled without causing side effects (in isolation).

**Comment by stof** [comment]
Rebooting the kernel in the messenger consumer is not possible. We would end up with an invalid state as the consumer itself is manager by the container of the kernel. Rebooting a kernel can only be done by code living *outside* the kernel-managed application, not from inside it. That's precisely why we have the service resetter which is about resetting stateful services for a long-running process.
Resetting the container in such place will not work either, as the worker consumer itself is managed by the container (same reason of not working than for a full reboot, as both a full reboot and a container reset are invalidating all service instances).
IF the service resetter was resetting the container itself (instead of only resetting stateful services), we would not be able to use it to reset state between messenger messages.

Your case of multi-tenancy should be handled using a stateful service from which other services can read that state (not in their constructor but when they need to be used instead) instead of using a custom env var loader to change the values injected in their constructor during instantiation if you want to be compatible with long-running processes that don't re-instantiate services.

I put a big -1 on the proposed "fix" as it would break messenger for all projects.

**Comment by faizanakram99** [comment]
> Resetting the container in such place will not work either, as the worker consumer itself is managed by the container (same reason of not working than for a full reboot, as both a full reboot and a container reset are invalidating all service instances).

Good point, the reset can only reset env vars, it doesn't have to reset all the services. 

> Your case of multi-tenancy should be handled using a stateful service from which other services can read that state (not in their constructor but when they need to be used instead)

How'll doctrine work in such a scenario, particularly separate database per tenant?
I've edited the OP to make it a bit more clear.

Same goes for things like Filesystem (flysystem), cache, etc.

**Comment by faizanakram99** [comment]
@stof Thank you for pointing the flaw in the previous suggested fix, I've added a PR (#59136) with least distrubtive changes
